site/cds_rdm/tasks.py

```diff
@@ -140,3 +147,224 @@ def sync_users():
+                    _update_name(user, name_dict, is_orcid=True)
+                except NoResultFound:
+                    _create_new_name(user)
+            else:
+                _create_new_name(user)
+
+
+
+@shared_task()
+def merge_duplicate_names_vocabulary(since=None):
+    """Merges duplicate names in the names vocabulary."""
+    service = current_service_registry.get("names")
+
+    def _merge(name_from, name_to):
+        """Merges the names.
+
+        param name_from: The name to merge from. This will marked as unlisted after the merge, meaning that won't be returned in the search results.
+        param name_to: The name to merge to.
+        """
+        updated = False
+        # Merge the affiliations
+        affiliations_to = name_to.get("affiliations", [])
+        for affiliation_from in name_from.get("affiliations", []):
+            if affiliation_from not in affiliations_to:
+                affiliations_to.append(affiliation_from)
+                updated = True
+
+        name_to["affiliations"] = affiliations_to
+
+        # Merge the identifiers
+        identifiers_to = name_to.get("identifiers", [])
+        for identifier_from in name_from.get("identifiers", []):
+            if identifier_from not in identifiers_to:
+                identifiers_to.append(identifier_from)
+                updated = True
+        name_to["identifiers"] = identifiers_to
+
+        # Merge the tags (except "unlisted" value)
+        tags_to = name_to.get("tags", [])
+        tags_from = name_from.get("tags", [])
+        for tag_from in tags_from:
+            if tag_from not in tags_to and tag_from != "unlisted":
+                tags_to.append(tag_from)
+                updated = True
+        name_to["tags"] = tags_to
+
+        # Merge the props
+        props_to = name_to.get("props", {})
+        for key, value in name_from.get("props", {}).items():
+            if key not in props_to:
+                props_to[key] = value
+                updated = True
+        name_to["props"] = props_to
+
+        # Mark as unlisted the name_from value
+        if "unlisted" not in tags_from:
+            if tags_from:
+                name_from["tags"] = tags_from + ["unlisted"]
+            else:
+                name_from["tags"] = ["unlisted"]
+
+            # "Deprecate" the name_from value
+            service.update(system_identity, name_from.get("id"), name_from)
+
+        if updated:
+            # Update the name_to value
+            service.update(system_identity, name_to.get("id"), name_to)
+
+    filters = [
+        dsl.Q("term", **{"props.is_cern": True}),
+        dsl.Q("term", **{"identifiers.scheme": "orcid"}),
+        dsl.Q("bool", must_not=[dsl.Q("term", tags="unlisted")]),
+    ]
+    if since:
+        filters.append(dsl.Q("range", updated={"gte": since}))
+    combined_filter = dsl.Q("bool", filter=filters)
+
+    names = service.scan(system_identity, extra_filter=combined_filter)
```

Comment:
shouldn't we rely on the db table instead?

Reply:
Given that we are applying a lot of filters, plus there depending on the time range and/or when we just imported ORCID and the CERN authors, there might be a lot of results to process and there are millions of values in this table, I would expect this to be quite slow.

I am aware that the DB is the source of truth but mainly for performance optimization I would go with OS, however, if there are strong reasons to prioritize the DB, I can change it and test it with a bigger sample to asses the performance